### DISCUSSION TITLE
why can't I create a single table when using asyncio?  AttributeError: 'AsyncEngine' object has no attribute '_run_ddl_visitor'

### DISCUSSION THREAD

**Discussion by iamthebull** [sqlalchemy/sqlalchemy#11998] [Usage Questions] [answered]
why can't I use:
 ```
table.create()
```
 or
 ```
async with db.engine.begin() as conn:
	await conn.run_sync(partial(db.meta.create_all,db.engine,[table]))
```
When using asyncio? I get the following error?
```
AttributeError: 'AsyncEngine' object has no attribute '_run_ddl_visitor'
```
A sample program:
```
python 3.12
sqlalchemy 2.0.35
```
```
import asyncio
import logging
from sqlalchemy import MetaData, Table, Column, Integer, String
from sqlalchemy.ext.asyncio import create_async_engine
from functools import partial
import logging

logger = logging.getLogger()

class DB(): pass
db = DB()

class Reg():
	def __init__(self, name) -> None:
		self.name = name
		self.table = self.getTable()
		# self.table.create(db.engine)    <---- this doesn't work

	def getTable(self) -> Table:
		return Table(
			'items',
			db.meta,
			Column('id', Integer, primary_key=True),
			Column('name', String(30)),
			Column('type', String(15))
		)
	
	async def createTable(self, table:Table):
		async with db.engine.begin() as conn:
			await conn.run_sync(db.meta.create_all)      # <------   this works
			# await conn.run_sync(partial(db.meta.create_all,db.engine,[table]))    <----- this doesn't work

async def main():
	logging.basicConfig(level=logging.INFO)

	db.engine = create_async_engine(f'sqlite+aiosqlite:///:memory:', echo=True)
	db.meta = MetaData()

	reg = Reg('Reg')
	await reg.createTable(reg.table)

	await db.engine.dispose()

if __name__ == '__main__':
	asyncio.run(main())
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

It's currently not possible, since create_all cannot currently create an async connection.

You will need to use, as mentioned in your example:
```py
async with db.engine.begin() as conn:
    await conn.run_sync(db.meta.create_all)
```

**Reply by iamthebull** [reply]
Ok thanks. Do you know if there are plans to make an async version of this function in the future?

**Reply by CaselIT** [reply]
Not at the moment, since currently the only parts that are async are in the `ext.asyncio` module.

I wouldn't be opposed to add async utility methods to select items, something like (it would actually need to support `AsyncConnection` too)
```py
class MetaData:
  ...
  async def create_all_async(self, bind: AsyncEngine):
    async with bind.begin() as conn:
      await conn.run_sync(self.create_all)
```

Maybe something for that could be added for 2.1.
@zzzeek what do you think about it?

**Reply by gordthompson** [reply]
FWIW, Stack Overflow answer [here](https://stackoverflow.com/a/79086460/2144390) says that a lambda would work to specify a single table.

**Reply by CaselIT** [reply]
if the only request is to pass a single table, just pass it as arguments to run_sync. It's even typed correctly to tell you what args are accepted. Just use `await conn.run_sync(metadata.create_all, [table])`

**Reply by iamthebull** [reply]
Yes, that was the original request. And your suggestion worked thanks. I was also just curious if `create_all` would always be a blocking sync function or if it would be optimized for asyncio at some point.

**Reply by CaselIT** [reply]
since you can't make a single function both sync and async (with a simple api) at best we are looking at an async alias, like I mentioned above

**Reply by iamthebull** [reply]
I thought `create_all` was a blocking function. Meaning that although it is running inside the asyncio event loop it would be blocking while it was being executed. So my question was whether another separate function would be provided that did the same thing (create all the tables) without blocking.

It seems my understanding of the `create_all` function was incorrect. Information in [this post](https://github.com/sqlalchemy/sqlalchemy/discussions/5914#discussioncomment-336507) made it more clear.

> un_sync() does not accept a coroutine, is accepts a plain Python function which calls blocking style API functions that ultimately call awaitables for the actual IO. it's important to note the difference in terminology between "blocking style" and "blocking". SQLAlchemy's asyncio extension does not run "blocking" IO calls. The run_sync() method runs code that is written in "blocking style" but ultimately is using asyncio using greenlets.

`create_all` is not blocking so there is no need for a separate function as I was originally requesting. It's all good. Thanks.

**Reply by CaselIT** [reply]
Ok, in any cade it's a valid request, even of it's just a wrapper around the existing create_all.

Let's wait @zzzeek opinion

**Reply by zzzeek** [reply]
what's the request?   create_all and any other function use the asyncio driver and are non blocking.

**Reply by CaselIT** [reply]
Add utility function on metadata (and schema items too?) that can be used directly with async engine/connection. example: https://github.com/sqlalchemy/sqlalchemy/discussions/11998#discussioncomment-10937247

**Reply by zzzeek** [reply]
that's a slope that will keep slipping IMO.   I'm really not sold on putting async methods throughout the main Core, it seems like this could lead to bigger contradictions in the API we aren't going to want

**Reply by CaselIT** [reply]
I think it could work if we pick a standard naming convention, and only as syntactic sugar over an existing method that accepts a bind.
Like if we have a `method_with_bind(bind, ...)` there could be a `async_method_with_bind(bind, ...)` or `method_with_bind_async(bind, ...)` that calls the original method with using `run_sync`. From a quick look the methods seem to be 5-6

Another alternative is to overload the `method_with_bind` to return a coroutine if the bind is async. The method would keep the plain `def ` declaration, just return a T/coroutine[T] (where T seems to be none almost in all cases).
